### ISSUE TITLE
Memory leak with kafka metrics

### ISSUE THREAD

**Issue by gstephant** [micrometer-metrics/micrometer#1743]
Hi there,

I have a springboot app wich purpose is to print the lag on all kafka topics/consumer groups. And that's all.

It runs out of memory, and the memory usage is as follows :

![image](https://user-images.githubusercontent.com/22791611/70515272-5ca8bb80-1b35-11ea-9537-c33bedd82519.png)

Code : 

```java
@Service
@ManagedResource
public class KafkaLagWriter {

    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaLagWriter.class);

    private final KafkaAdmin kafkaAdmin;
    private final String bootstrapServers;

    public KafkaLagWriter(final KafkaAdmin kafkaAdmin,
                          @Value("${spring.kafka.bootstrap-servers}") final String bootstrapServers) {
        this.kafkaAdmin = kafkaAdmin;
        this.bootstrapServers = bootstrapServers;
    }

    @ManagedOperation
    @Scheduled(fixedDelay = 60_000)
    public void logKafkaTopicLags() {

        try (final AdminClient adminClient = AdminClient.create(kafkaAdmin.getConfig())) {
            final ListConsumerGroupsResult consumerGroupsResult = adminClient.listConsumerGroups();
            final Collection<ConsumerGroupListing> consumerGroupListing = consumerGroupsResult.all().get();
            for (final ConsumerGroupListing consumerGroup : consumerGroupListing) {
                final String consumerGroupId = consumerGroup.groupId();

                final ListConsumerGroupOffsetsResult consumerGroupOffsetsResult = adminClient.listConsumerGroupOffsets(consumerGroup.groupId());
                final KafkaFuture<Map<TopicPartition, OffsetAndMetadata>> futureMap = consumerGroupOffsetsResult.partitionsToOffsetAndMetadata();
                final Map<TopicPartition, OffsetAndMetadata> offSetByPartition = futureMap.get();
                final Map<String, Map<TopicPartition, OffsetAndMetadata>> topicsToOffSetByPartition = new HashMap<>();

                offSetByPartition.forEach((key, value) -> {
                    Map<TopicPartition, OffsetAndMetadata> offSets = topicsToOffSetByPartition.get(key.topic());
                    if (offSets != null) {
                        offSets.put(key, value);
                    } else {
                        offSets = new HashMap<>();
                        offSets.put(key, value);
                    }
                    topicsToOffSetByPartition.put(key.topic(), offSets);
                });

                for (final Entry<String, Map<TopicPartition, OffsetAndMetadata>> topicToPartitions : topicsToOffSetByPartition.entrySet()) {
                    logOffSets(consumerGroupId, topicToPartitions.getKey(), topicToPartitions.getValue());
                }
            }
        } catch (final Exception e) {
            final ExceptionEvent event = ExceptionEventBuilder.createBuilder()
                                                              .method(Events.KAFKA_MONITORING)
                                                              .exceptionClassName(e.getClass().getName())
                                                              .exceptionMessage(e.getMessage())
                                                              .stacktrace(ExceptionUtils.getStackTrace(e))
                                                              .build();
            LOGGER.error(event, null);
        }
    }

    private void logOffSets(final String groupId, final String topic, final Map<TopicPartition, OffsetAndMetadata> offSetByPartition) {
        try (final KafkaConsumer<?, ?> consumer = createNewConsumer(groupId)) {
            final Map<TopicPartition, Long> endOffSetByPartition = consumer.endOffsets(offSetByPartition.keySet());
            for (final TopicPartition topicPartition : offSetByPartition.keySet()) {
                final Long endOffSet = endOffSetByPartition.get(topicPartition);
                final long currentOffSet = offSetByPartition.get(topicPartition).offset();
                final Long lag = endOffSet != null ? endOffSet - currentOffSet : null;

                final KafkaConsumerStatisticEvent event = KafkaConsumerStatisticEventBuilder.createBuilder()
                                                                                            .topic(topic)
                                                                                            .consumerGroup(groupId)
                                                                                            .partition(topicPartition.partition())
                                                                                            .currentOffSet(currentOffSet)
                                                                                            .endOffSet(endOffSet)
                                                                                            .lag(lag)
                                                                                            .build();
                LOGGER.info(event, null);
            }
        }
    }

    private KafkaConsumer<?, ?> createNewConsumer(final String groupId) {
        final Properties properties = new Properties();
        properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        properties.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
        properties.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
        properties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
        properties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
        return new KafkaConsumer<>(properties);
    }
}
```

Can you provide some help on this matter ?

    <micrometer.version>1.1.4</micrometer.version>

**Comment by checketts** [comment]
I haven't used the Kafka metrics, however can you tell how many meters had been created? Since those are held in memory, typically a high cardinality tag is to blame.

**Comment by gstephant** [comment]
226226 objects Meter.Id
![image](https://user-images.githubusercontent.com/22791611/70534093-9c35ce80-1b5a-11ea-89a5-0d58eeaae74b.png)

**Comment by checketts** [comment]
Are you not registering any of those meters yourself? Are all those from the Kafka meter binder? Could you still into them and see what specific tag is causing the high cardinality? Adding the LoggingMeterRegistry will write them to logs. 

A quick fix for you would be to add a MeterFilter to keep them count in check. But any additional info you can give around the specifics will help us fix it if it is a bug in the binder.

**Comment by gstephant** [comment]
I dont register anything myself : thats my pain here, i dont need these metrics, and they keep increasing over time.
My guess is each time I create a consumer object, lot of meters are initiated in the background, and when the consumer is closed, those meters are not released.

**Comment by checketts** [comment]
Sounds right. I'll give you a snippet that will combine all the Kafka consumer meters into a single one once I get to my computer in a few hours.

**Comment by izeye** [comment]
> i dont need these metrics

@gstephant If you don't need them, you can disable them with the following application property:

```properties
spring.autoconfigure.exclude=org.springframework.boot.actuate.autoconfigure.metrics.KafkaMetricsAutoConfiguration
```

**Comment by izeye** [comment]
I created https://github.com/micrometer-metrics/micrometer/pull/1745 to try to resolve this.

**Comment by shakuzen** [comment]
@gstephant Would you please try the latest 1.1.10-SNAPSHOT version and let us know if the problem is fixed?

**Comment by jorgheymans** [comment]
@shakuzen was this fixed in micrometer-core 1.3.3 as well ? Coming here from https://github.com/tchiotludo/kafkahq/issues/209 , we're still seeing the memory leak. I can provide you with a memory dump in case this helps.

**Comment by shakuzen** [comment]
@jorgheymans The fix was included in 1.3.3 as well. If you're still seeing an issue, could you open a separate issue with details so we can take a look and hopefully fix it? Also of note is that we have a non-JMX-based Kafka metrics option in the upcoming 1.4 release that intends to replace the JMX-based Kafka metrics. see #1835

**Comment by alevohin** [comment]
@jorgheymans @shakuzen I suppose the problem with memory leak is described here #2096.
Version 1.3.X is still used by Springboot 2.2.6.RELEASE
